node/impl/full/mpool.go

```diff
@@ -67,36 +67,25 @@ func (a *MpoolAPI) MpoolPending(ctx context.Context, tsk types.TipSetKey) ([]*ty
 		return nil, xerrors.Errorf("loading tipset %s: %w", tsk, err)
 	}
 	pending, mpts := a.Mpool.Pending()
+	if ts == nil || mpts.Height() > ts.Height() {
+		return pending, nil
+	}
 
 	haveCids := map[cid.Cid]struct{}{}
 	for _, m := range pending {
 		haveCids[m.Cid()] = struct{}{}
 	}
 
-	if ts == nil || mpts.Height() > ts.Height() {
-		return pending, nil
-	}
-
 	for {
 		if mpts.Height() == ts.Height() {
 			if mpts.Equals(ts) {
 				return pending, nil
 			}
-			// different blocks in tipsets
-
-			have, err := a.Mpool.MessagesForBlocks(ts.Blocks())
-			if err != nil {
-				return nil, xerrors.Errorf("getting messages for base ts: %w", err)
-			}
-
-			for _, m := range have {
-				haveCids[m.Cid()] = struct{}{}
-			}
```

Comment:
I am not convinced that this is a bug at all.

The current code sees that we have different blocks in the tipset and tries to select pending messages that were not included in the previous (known) tipset at the same height.

With this change, we will select pending messages that were included in the previous tipset, even though some other block may contain them.

I think that this change actually introduces a bug instead of fixing one.

Reply:
we should probably discuss this in the standup as it is a very subtle change with potentially nasty effects. For now, I am inclined to say no.